bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/items/ItemStateConverterImpl.java

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright (c) 2014,2017 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.internal.items;
+
+import javax.measure.Quantity;
+import javax.measure.Unit;
+import javax.measure.quantity.Dimensionless;
+
+import org.eclipse.smarthome.core.i18n.UnitProvider;
+import org.eclipse.smarthome.core.items.Item;
+import org.eclipse.smarthome.core.items.ItemStateConverter;
+import org.eclipse.smarthome.core.library.items.NumberItem;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.types.MeasurementSystem;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.StateDescription;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.eclipse.smarthome.core.types.util.UnitUtils;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Convert a {@link State} to an {@link Item} accepted {@link State}.
+ *
+ * @author Henning Treu - initial refactoring as OSGi service
+ *
+ */
+@Component(service = ItemStateConverter.class)
+public class ItemStateConverterImpl implements ItemStateConverter {
+
+    private final Logger logger = LoggerFactory.getLogger(ItemStateConverterImpl.class);
+
+    private UnitProvider unitProvider;
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @Override
+    public State convertToAcceptedState(State state, Item item) {
+        if (state == null) {
+            logger.error("A conversion of null was requested",
+                    new IllegalArgumentException("state should not be null"));
+            return UnDefType.NULL;
```

Comment:
Since ItemStateConverter is a consumer interface, would you rate the `@NonNull` parameter annotation and IAE higher then accepting `@Nullable State state` and returning UndefType.NULL instead?

Reply:
My initial refactoring was the latter, so drop the error log completely and accept nullable state. Just convert it and document the behaviour in the JavaDoc. wdyt?

Reply:
Consumer interface does not mean that null parameters are always valid and accepted. If it is mandatory as in this case, we should still mark it as NonNull, but nonetheless do a null check in our implementation and throw an IAE, if we receive null. Imho that would be a suitable general convention for us: Within the framework code, throw IAEs whenever a required (i.e. NonNull) parameter is passed in as null.